csharp/rust/src/lib.rs

```diff
@@ -1,284 +1,472 @@
-
-impl From<logger_core::Level> for Level {
-    fn from(level: logger_core::Level) -> Self {
-        match level {
-            logger_core::Level::Error => Level::Error,
-            logger_core::Level::Warn => Level::Warn,
-            logger_core::Level::Info => Level::Info,
-            logger_core::Level::Debug => Level::Debug,
-            logger_core::Level::Trace => Level::Trace,
-            logger_core::Level::Off => Level::Off,
-        }
     }
-}
-
-impl From<Level> for logger_core::Level {
-    fn from(level: Level) -> logger_core::Level {
-        match level {
-            Level::Error => logger_core::Level::Error,
-            Level::Warn => logger_core::Level::Warn,
-            Level::Info => logger_core::Level::Info,
-            Level::Debug => logger_core::Level::Debug,
-            Level::Trace => logger_core::Level::Trace,
-            Level::Off => logger_core::Level::Off,
-        }
+    CreateClientHandleResult {
+        result: ECreateClientHandleCode::Success,
+        client_handle: Box::into_raw(Box::new(FFIHandle { runtime, handle })) as *const c_void,
+        error_string: null(),
     }
 }
 
-/// Unsafe function because creating string from pointer.
+/// # Summary
+/// Frees the previously created client_handle, making it unusable.
+///
+/// # Input Safety (in_...)
+/// The data passed in is considered "caller responsibility".
+/// Any pointers hence will be left unreleased after leaving.
 ///
-/// # Safety
+/// # Output Safety (out_... / return ...)
+/// The data returned is considered "caller responsibility".
+/// The caller must release any non-null pointers.
 ///
-/// * `message` and `log_identifier` must not be `null`.
-/// * `message` and `log_identifier` must be able to be safely casted to a valid [`CStr`] via [`CStr::from_ptr`]. See the safety documentation of [`std::ffi::CStr::from_ptr`].
+/// # Reference Safety (ref_...)
+/// Any reference data is considered "caller owned".
+///
+/// # Freeing data allocated by the API
+/// To free data returned by the API, use the corresponding `free_...` methods of the API.
+/// It is **not optional** to call them to free data allocated by the API!
 #[no_mangle]
-#[allow(improper_ctypes_definitions)]
-pub unsafe extern "C" fn log(
-    log_level: Level,
-    log_identifier: *const c_char,
-    message: *const c_char,
-) {
-    unsafe {
-        logger_core::log(
-            log_level.into(),
-            CStr::from_ptr(log_identifier)
-                .to_str()
-                .expect("Can not read log_identifier argument."),
-            CStr::from_ptr(message)
-                .to_str()
-                .expect("Can not read message argument."),
-        );
-    }
+pub extern "C-unwind" fn csharp_free_client_handle(in_client_ptr: *const c_void) {
+    logger_core::log(Trace, "csharp_ffi", "Entered csharp_free_client_handle");
+    let client_ptr = unsafe { Box::from_raw(in_client_ptr as *mut FFIHandle) };
+    let _runtime_handle = client_ptr.runtime.enter();
```

Comment:
No need to do anything with runtime, but please keep this
```suggestion
    unsafe { Arc::decrement_strong_count(client_ptr as *const Client) };
```